enhancements/single-node-production-deployment-approach.md

```diff
@@ -0,0 +1,484 @@
+potentially be drained, but there is no way for the MCO to know if the
+workers are fully utilized. When the high-availability mode is `None`
+the step needs to be skipped. Ideally we will be able to rely on the
+[graceful shutdown support in
+kubelet](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2000-graceful-node-shutdown)
+to stop all workloads safely as part of the reboot.
+
+#### cluster-authentication-operator
+
+By default, the `cluster-authentication-operator` will not deploy
+oauth-apiserver and oauth-server without a minimum of 3 control plane
+nodes. This can be change by enabling
+`useUnsupportedUnsafeNonHANonProductionUnstableOAuthServer`. The
+operator will need to be updated to honor the new capabilities API
+instead. Health checks for these components will also have to be
+updated to allow a single replica.
+
+#### cluster-etcd-operator
+
+By default, `cluster-etcd-operator` will not deploy the etcd cluster
+without minimum of 3 master nodes. This can be changed by enabling
+`useUnsupportedUnsafeNonHANonProductionUnstableEtcd`. The operator
+will need to be updated to honor the new capabilities API instead.
+
+`etcd-quorum-guard` is managed by the cluster-version-operator, which
+means it has a statically configured Deployment. The operator will
+need to be updated to manage the `etcd-quorum-guard` so that it can
+support changing the replica count or disabling the guard entirely,
+depending on the best approach to honor the high-availability mode
+setting in the capabilities API.
+
+#### cluster-ingress-operator
+
+By default, the `cluster-ingress-operator` deploys the router with 2
+replicas. On a single node one will fail to start and the ingress will
+show as degraded. The operator will need to be updated to honor the
+new capabilities API to change the replica count to 1 when the
+high-availability mode is none.
+
+#### cluster-machine-approver
+
+The operator should disable automatic machine approval. (See risks
+section below.)
+
+#### operator-lifecycle-manager
+
+Operators managed by the `operator-lifecycle-manager` will look at the
+`infrastructure` API if they need to configure their operand
+differently based on the high-availability mode.
+
+### Risks and Mitigations
+
+*How will security be reviewed and by whom? How will UX be reviewed and by whom?*
+
+#### Control plane rollouts
+
+kube-apiserver is periodically reconfigured by its operator in
+response to changes in cluster state (e.g. weekly rotation of etcd
+encryption keys). When this occurs, the apiserver may be inaccessible
+for up to 2 minutes. Workloads that do not depend on the apiserver
+would be expected to run without issue during this interval. Workloads
+that do depend on apiserver availability would need to be resilient to
+these events. OpenShift core components are already resilient in this
+way.
+
+#### Cluster Machine Approver
+
+Auto-approval of certificate signing requests requires 2 sources of
```

Comment:
i am not sure it needs to be disabled.  the cluster just will never have a machine with a corresponding nodeRef, so it would be inert.

Reply:
When you say "machine" do you mean "Machine"? Because I think we will have a Machine resource, right?

Reply:
correct, i mean `Machine`, but the `Machine` resource you may have is already approved.  I do not anticipate you have a second `Machine` resource, so the approver could be on and inert.

Reply:
The cluster machine approver is also used for renewals of certificates in OCP, I'm not sure how long the certificates are valid for by default but if there's no `Machine` for the single node, it will just spin saying that it can't find the `Machine` for a `Node` when there is a CSR for a renewal to approve. Something else may be needed to handle the renewals that doesn't rely on the `Machine`

Reply:
I've tried to incorporate your comment into this section in https://github.com/openshift/enhancements/pull/596. Please let me know if I understood the implications correctly.